File: src/PcIfy.Client/Services/BookmarkService.cs
using System.Text.Json;
using PcIfy.Client.Models;
using PcIfy.Client.Services.Interfaces;

namespace PcIfy.Client.Services;

public class BookmarkService : IBookmarkService
{
    private const string PrefsKey = "bookmarks_json";
    private List<BookmarkedFolder> _bookmarks;

    public event EventHandler? BookmarksChanged;

    public BookmarkService()
    {
        var json = Preferences.Get(PrefsKey, null);
        _bookmarks = json is null ? [] : JsonSerializer.Deserialize<List<BookmarkedFolder>>(json) ?? [];
    }

    public IReadOnlyList<BookmarkedFolder> GetBookmarks() => _bookmarks.AsReadOnly();

    public void AddBookmark(string path, string displayName, string? coverPath = null)
    {
        _bookmarks.RemoveAll(b => b.Path == path);
        _bookmarks.Add(new BookmarkedFolder { Path = path, DisplayName = displayName, CoverThumbnailServerPath = coverPath });
        Persist();
    }

    public void RemoveBookmark(string path)
    {
        _bookmarks.RemoveAll(b => b.Path == path);
        Persist();
    }

    private void Persist()
    {
        Preferences.Set(PrefsKey, JsonSerializer.Serialize(_bookmarks));
        BookmarksChanged?.Invoke(this, EventArgs.Empty);
    }
}

File: src/PcIfy.Client/lib/services/bookmark_service.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

import '../core/models/bookmarked_folder.dart';

class BookmarkService {
  BookmarkService(this._prefs);

  final SharedPreferences _prefs;
  static const _key = 'bookmarks_json';

  List<BookmarkedFolder> getBookmarks() {
    final json = _prefs.getString(_key);
    if (json == null) return [];
    try {
      final list = jsonDecode(json) as List<dynamic>;
      return list
          .map((e) => BookmarkedFolder.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (_) {
      return [];
    }
  }

  Future<void> addBookmark(BookmarkedFolder bookmark) async {
    final list = getBookmarks();
    if (list.any((b) => b.path == bookmark.path)) return;
    list.add(bookmark);
    await _save(list);
  }

  Future<void> removeBookmark(String path) async {
    final list = getBookmarks()..removeWhere((b) => b.path == path);
    await _save(list);
  }

  Future<void> reorder(List<BookmarkedFolder> newOrder) => _save(newOrder);

  bool isBookmarked(String path) => getBookmarks().any((b) => b.path == path);

  Future<void> _save(List<BookmarkedFolder> list) =>
      _prefs.setString(_key, jsonEncode(list.map((b) => b.toJson()).toList()));
}
